routes/cube_routes.js

```diff
@@ -1248,103 +1159,59 @@ router.post('/bulkuploadfile/:id', ensureAuth, async (req, res) => {
   }
 });
 
-async function bulkuploadCSV(req, res, cards, cube) {
-  let added = [];
-  let missing = '';
-  let changelog = '';
-  for (let card_raw of cards) {
-    let split = util.CSVtoArray(card_raw);
-    let name = split[0];
-    let maybeboard = split[8];
-    let card = {
-      name: name,
-      cmc: split[1],
-      type_line: split[2].replace('-', '—'),
-      colors: split[3].split('').filter((c) => [...'WUBRG'].includes(c)),
-      set: split[4].toUpperCase(),
-      addedTmsp: new Date(),
-      collector_number: split[5],
-      status: split[6],
-      finish: split[7],
-      imgUrl: split[9] && split[9] != 'undefined' ? split[9] : null,
-      tags: split[10] && split[10].length > 0 ? split[10].split(',') : [],
-    };
+router.post('/bulkreplacefile/:id', ensureAuth, async function(req, res) {
+  if (!req.files) {
+    req.flash('danger', 'Please attach a file');
+    return res.redirect('/cube/list/' + req.params.id);
+  }
+
+  const items = req.files.document.data.toString('utf8'); // the uploaded file object
 
-    let potentialIds = carddb.allIds(card);
-    if (potentialIds && potentialIds.length > 0) {
-      // First, try to find the correct set.
-      let matchingSet = potentialIds.find((id) => carddb.cardFromId(id).set.toUpperCase() == card.set);
-      let nonPromo = potentialIds.find(carddb.notPromoOrDigitalId);
-      let first = potentialIds[0];
-      card.cardID = matchingSet || nonPromo || first;
-      if (maybeboard == 'true') {
-        cube.maybe.push(card);
+  try {
+    const cube = await Cube.findOne(build_id_query(req.params.id));
+    if (!cube) {
+      req.flash('danger', 'Cube not found');
+      return res.status(404).render('misc/404', {});
+    } else if (cube.owner != req.user._id) {
+      req.flash('danger', 'Not Authorized');
+      return res.redirect('/cube/list/' + req.params.id);
+    }
+    cards = items.match(/[^\r\n]+/g);
+    if (cards) {
+      if (cards[0].trim() == CSV_HEADER) {
+        cards.splice(0, 1);
+        const { newCards, newMaybe, missing } = CSVtoCards(cards, carddb);
+        await populateCardDetails([cube.cards, newCards, newMaybe], carddb, prices);
+        const { only_a, only_b, in_both, all_cards } = await compareCubes(cube.cards, newCards);
+        let changelog = '';
+        only_a.forEach((card) => (changelog += removeCardHtml(carddb.cardFromId(card.cardID))));
```

Comment:
We should put this in a style guide or Contributing since it seems forEach was being used throughout much of the code which led to me copying it to use when I added things.

Reply:
This is something Patrick had confronted me about in person last weekend, so we decided to enforce it moving forward. Not a big deal either way as it's an easy change